File: huawei/HJ16.cpp
// 购物单

#include <iostream>
#include <vector>

using std::max;
using std::vector;

#define MAIN_ITEM 0
#define SCALE 10

struct Attachment
{
    Attachment(int price, int importance, int main_item)
        : price_(price), importance_(importance), main_item_(main_item)
    {
        satis_ = price_ * importance_;
    }
    int price_;
    int importance_;
    int main_item_;
    int satis_;
};

struct Item
{
    Item(int price, int importance, int index)
        : price_(price), importance_(importance), index_(index)
    {
        satis_ = price_ * importance_;
    }
    int                     price_;
    int                     importance_;
    int                     index_;
    int                     satis_;
    std::vector<Attachment> atts_;
};

int main()
{
    int budget_cap;
    int input_line_amount;
    std::cin >> budget_cap >> input_line_amount;

    std::vector<Item>       items;
    std::vector<Attachment> attachments;

    for (int i = 0; i < input_line_amount; i++)
    {
        int price;
        int importance;
        int main_item;
        std::cin >> price >> importance >> main_item;
        if (main_item == MAIN_ITEM)
            items.push_back(Item{price / SCALE, importance, i});
        else
            attachments.push_back(
                Attachment{price / SCALE, importance, main_item});
    }

    for (const auto& attachment : attachments)
    {
        for (auto& item : items)
        {
            if (item.index_ + 1 == attachment.main_item_)
                item.atts_.push_back(attachment);
        }
    }

    budget_cap /= SCALE;
    vector<vector<int>> satis(items.size() + 1, vector<int>(budget_cap + 1, 0));
    for (int i = 1; i < items.size() + 1; ++i)
    {
        const Item& item = items[i - 1];
        for (int budget = 1; budget < budget_cap + 1; ++budget)
        {
            // Option 0: Do not but this item
            satis[i][budget] = satis[i - 1][budget];

            // Option 1: But the main item alone
            if (item.price_ <= budget)
            {
                satis[i][budget] = max(satis[i][budget],
                                       satis[i - 1][budget - item.price_]
                                           + item.satis_);
            }

            // Option 2: Buy the main item and first attachment (if exists)
            if (item.atts_.size() >= 1
                && item.price_ + item.atts_[0].price_ <= budget)
            {
                satis[i][budget] = max(
                    satis[i][budget],
                    satis[i - 1][budget - item.price_ - item.atts_[0].price_]
                        + item.satis_ + item.atts_[0].satis_);
            }

            // Option 3: Buy the main item and second attachment (if exists)
            if (item.atts_.size() >= 2
                && item.price_ + item.atts_[1].price_ <= budget)
            {
                satis[i][budget] = max(
                    satis[i][budget],
                    satis[i - 1][budget - item.price_ - item.atts_[1].price_]
                        + item.satis_ + item.atts_[1].satis_);
            }

            // Option 4: Buy the main item with both attachments (if exist)
            if (item.atts_.size() >= 2
                && item.price_ + item.atts_[0].price_ + item.atts_[1].price_
                       <= budget)
            {
                satis[i][budget] = max(
                    satis[i][budget],
                    satis[i - 1][budget - item.price_ - item.atts_[0].price_
                                 - item.atts_[1].price_]
                        + item.satis_ + item.atts_[0].satis_
                        + item.atts_[1].satis_);
            }
        }
    }
    std::cout << satis[items.size()][budget_cap] * SCALE << "\n";
}

File: huawei/HJ16.py
from typing import List

SCALE = 10

class Att:
    price: int
    satis: int
    main_item_index: int
    def __init__(self, price: int, importance: int, main_item_index: int):
        self.price = price // SCALE
        self.satis = price * importance
        self.main_item_index = main_item_index


class Item:
    price: int
    satis: int
    index: int
    atts: List[Att]
    def __init__(self, price: int, importance: int, index: int):
        self.price = price // SCALE
        self.satis = price * importance
        self.index = index
        self.atts = []


def main():
    total_budget, input_lines = map(int, input().split(' '))
    items: List[Item] = []
    atts: List[Att] = []
    for index in range(input_lines):
        price, importance, main_item_number = map(int, input().split(' '))
        if main_item_number == 0:
            items.append(Item(price, importance, index))
        else:
            atts.append(Att(price, importance, main_item_number - 1))

    for att in atts:
        for item in items:
            if att.main_item_index == item.index:
                item.atts.append(att)
                break


    total_budget //= SCALE
    satis = [[0 for _ in range(total_budget + 1)] for _ in range (len(items)+1)]
    for i in range(1, len(items) + 1):
        item = items[i-1]
        for budget in range(1, total_budget + 1):
            # Option 0: Do not buy item
            satis[i][budget] = satis[i-1][budget]
            # Option 1: Buy item alone
            if item.price <= budget:
                satis[i][budget] = max(
                    satis[i][budget],
                    satis[i-1][budget-item.price] + item.satis
                )
            # Option 2: Buy item with first attachment
            if len(item.atts) >= 1 \
                and item.atts[0].price + item.price <= budget:
                satis[i][budget] = max(
                    satis[i][budget],
                    satis[i-1][budget-item.price-item.atts[0].price] \
                        + item.satis + item.atts[0].satis
                )
            # Option 3: Buy item with second attachment
            if len(item.atts) >= 2 \
                and item.atts[1].price + item.price <= budget:
                satis[i][budget] = max(
                    satis[i][budget],
                    satis[i-1][budget-item.price-item.atts[1].price] \
                        + item.satis + item.atts[1].satis
                )
            # Option 4: Buy item with with both attachments
            if len(item.atts) >= 2 \
                and item.atts[0].price + item.atts[1].price + item.price \
                    <= budget:
                satis[i][budget] = max(
                    satis[i][budget],
                    satis[i-1][budget-item.price-item.atts[0].price-item.atts[1].price] \
                        + item.satis + item.atts[0].satis + item.atts[1].satis
                )

    print(satis[-1][-1])

if __name__ == "__main__":
    main()

